### PR TITLE
QAI-191, 192, 193, 194, 195, 197: ETL Pipeline for BBC News

### PR DESCRIPTION
**Pull request by karthikstar**
For the following user stories:
QAI-191, QAI-192, QAI-193, QAI-194, QAI-195, QAI-197
- Collection with minimum viable schema implemented
- Successful retrieval of all data rows from BBC News API for all months from 2017-01 to 2024-09 (total of ~125k rows)
- Successful insertion of all raw data into a MongoDB collection
- Transformation pipeline which filters down to relevant columns only and retain rows with non-missing values for title, publishedDate, description, content
- Successful loading of transformed rows back into another MongoDB Collection
- Implemented a GET endpoint on middleware which returns users all the transformed BBC articles. This is facilitated by communication between ETL and Middleware via REST.

For reference:
Pre-transformation rows loaded in: 124750
Post-Transformation rows loaded into DB: 124731

### CODE REVIEW

**Comment by karthikstar** [comment]
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news/fetch  

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/transform_bbc/run

**Comment by karthikstar** [comment]
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/transform_bbc/run

To retrieve all transformed BBC news from database (within ETL):
curl -X GET http://localhost:10070/api/transformed-bbc-news/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/bbcnews/transformed

File: build.gradle [hunk 1]
```diff
@@ -38,7 +38,12 @@ repositories {
 }
 
 dependencies {
+    implementation("org.springframework.boot:spring-boot-starter-data-mongodb")
     implementation("org.springframework.boot:spring-boot-starter-web")
+    implementation("org.springframework.boot:spring-boot-starter-security")
+    implementation("org.springframework.boot:spring-boot-configuration-processor")
+    implementation("org.springframework.boot:spring-boot-starter-web")
```
Comment (Ivsucram) [inline build.gradle]: "implementation("org.springframework.boot:spring-boot-starter-web")" is being imported twice.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package sg.com.quantai.etl.controllers
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt:1]: Add a space (empty line) between the package and imports, making it easy to visualize quickly.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package sg.com.quantai.etl.controllers
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt:1]: Add a space (empty line) between the package and imports, making it easy to visualize quickly.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.web.bind.annotation.*
+import sg.com.quantai.etl.services.BbcNewsTransformationService
+import org.springframework.http.ResponseEntity
+import org.springframework.http.HttpStatus
+
+@RestController
+@RequestMapping("/transform_bbc")
+class BbcNewsTransformationController(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt:10]: Lines 9 and 10

The scope of the data is still the same: BBC News.
Therefore, a new controller is not necessary.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import org.springframework.http.HttpStatus
+import sg.com.quantai.etl.requests.BbcRawNewsArticleRequest
+import sg.com.quantai.etl.services.BbcNewsApiService
+import java.time.LocalDate
+
+@RestController
+@RequestMapping("/news")
+class BbcRawNewsController(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt]: Line 12:
This endpoint does not fully reflect the data flow. Use something like "news_headlines," and then organize the data flow via functions and endpoints.

Line 13:
The scope here is still News Headlines. There shouldn't be a Raw Controller, but only a News Headline Controller with specific endpoints and functions to handle the different data formats (raw, transformed) and origin (we have a single source at this moment: BBC).

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt (hunk 3 above)
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt]: Add a space (empty line) between the package and imports, making it easy to visualize quickly.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt (hunk 4 above)
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt]: Lines 9 and 10

The scope of the data is still the same: News Headlines.
Therefore, a new controller is not necessary.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt (hunk 2 above)
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt]: Add a space (empty line) between the package and imports, making it easy to visualize quickly.

File: src/main/kotlin/sg/com/quantai/etl/controllers/BbcTransformedNewsController.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import sg.com.quantai.etl.data.BbcTransformedNewsArticle
+import sg.com.quantai.etl.repositories.BbcTransformedNewsRepository
+
+@RestController
+@RequestMapping("/api/transformed-bbc-news")
+class BbcTransformedNewsController(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/controllers/BbcTransformedNewsController.kt]: Line 10:
This endpoint does not fully reflect the data flow. Use something like "news_headlines," then organize the data flow via functions and endpoints.

Line 11:
The scope here is still News Headlines. There shouldn't be a Transformed or Specific API Controller, but only a News Headline Controller with specific endpoints and functions to handle the different data formats (raw, transformed), origin (we have a single source at this moment: BBC), and behaviors (fetch from external, deliver to internal call, etc).

File: src/main/kotlin/sg/com/quantai/etl/data/BbcTransformedNewsArticle.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package sg.com.quantai.etl.data
+
+import org.springframework.data.annotation.Id
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.core.mapping.Document
+import java.time.LocalDate
+
+@Document(collection = "bbc_news_articles_transformed")
+data class BbcTransformedNewsArticle(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/data/BbcTransformedNewsArticle.kt]: The transformed data class should be an abstraction in the format necessary for the platform's operations. The platform (users, researchers, and internal calls) does not currently care about the origin of the transformed data. Therefore, this data class should be something like "news_headlines."

I know that we only have one data source: BBC News. But, as discussed before, if we have multiple data sources, we can use them to auto-complete each other's information and create a single transformed data record that fulfills the needs of our platform.

File: src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package sg.com.quantai.etl.data
+
+import org.bson.types.ObjectId
+import org.springframework.data.annotation.Id
+import org.springframework.data.mongodb.core.index.Indexed
+import org.springframework.data.mongodb.core.mapping.Document
+import java.time.LocalDate
+
+@Document(collection = "bbc_news_articles_raw")
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt]: If you call the other data collection "news_headlines," call this something like "news_headlines_BBC" to indicate that it is a news_headlines of type BBC (where BBC automatically denotes the origin of the raw data).

File: src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+data class BbcRawNewsArticle(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt]: If you call the other data class "NewsHeadlines," call this something like "NewsHeadlinesBBC" to indicate that this is NewsHeadliens of type BBC (where BBC automatically denotes the origin of the raw data).

File: src/main/kotlin/sg/com/quantai/etl/data/BbcTransformedNewsArticle.kt [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    val title: String,
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/data/BbcTransformedNewsArticle.kt]: As suggested via written message on Discord, please use the format specified at https://github.com/Quant-NTU/persistence-middleware/blob/main/src/main/kotlin/sg/com/quantai/middleware/data/NewStrategy.kt to organize your data class

File: src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    val title: String,
+    val publishedDate: LocalDate,
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt]: As suggested via written message on Discord, please use the format specified at https://github.com/Quant-NTU/persistence-middleware/blob/main/src/main/kotlin/sg/com/quantai/middleware/data/NewStrategy.kt to organize your data class

File: src/main/kotlin/sg/com/quantai/etl/repositories/BbcTransformedNewsRepository.kt [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package sg.com.quantai.etl.repositories
+
+import sg.com.quantai.etl.data.BbcTransformedNewsArticle
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/repositories/BbcTransformedNewsRepository.kt]: This class does not exist

File: src/main/kotlin/sg/com/quantai/etl/requests/BbcRawNewsArticleRequest.kt [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package sg.com.quantai.etl.requests
+
+class BbcRawNewsArticleRequest(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/requests/BbcRawNewsArticleRequest.kt]: Change the name of this class accordingly to the name of the refactored data class..

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+import java.time.YearMonth
+import reactor.util.retry.Retry
+import java.time.Duration
+
+@Service
+class BbcNewsApiService(
+    @Autowired private val bbcRawNewsRepository: BbcRawNewsRepository
+) {
+
+    private val newApiUrl = "https://datasets-server.huggingface.co/"
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: The url string should be in the properties file.
The variable name "newApiUrl" should be updated accordingly.

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.http.MediaType
+import java.time.LocalDate
+import java.time.YearMonth
+import reactor.util.retry.Retry
+import java.time.Duration
+
+@Service
+class BbcNewsApiService(
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: The class name should be updated accordingly, such as "NewsHeadlineBBCService"

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    @Autowired private val bbcRawNewsRepository: BbcRawNewsRepository
+) {
+
+    private val newApiUrl = "https://datasets-server.huggingface.co/"
+    private val batchSize = 100
+
+    private fun bbcNewsWebClient(): WebClient {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: the method name can be changed to just webClient without losing its meaning

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        return WebClient.builder().baseUrl(newApiUrl)
+            .codecs { it.defaultCodecs().maxInMemorySize(16 * 1024 * 1024) }
+            .build()
+    }
+
+    fun fetchAndSaveNewsArticlesForAllMonths() {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Refactor title to "fetchAndSaveFromAllMonths"

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        println("Loading of BBC News Articles - START")
+
+        val startDate = YearMonth.of(2017, 1)
+        val endDate = YearMonth.of(2024, 9)
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: End Date should be set programmatically, and not static

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+        var currentMonth = startDate
+        while (currentMonth <= endDate) {
+            val configDate = currentMonth.toString()
+            println("Fetching data for: $configDate")
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Avoid using print, println and others. Use logs instead.

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+            fetchAndSaveNewsArticlesForDate(configDate)
+            currentMonth = currentMonth.plusMonths(1)
+        }
+        println("Loading of BBC News Articles - END")
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Avoid using print, println and others. Use logs instead.

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    private fun fetchAndSaveNewsArticlesForDate(config: String) {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Refactor title to "fetchAndSaveFromDate."
Use a more descriptive data type for config, such as java.time
User a more descriptive name for the variable as well, as config is too generic and doesn't give any information that you are retrieving data related to a date range

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+            processRows(rootNode.path("rows"))
+
+            for (offset in batchSize until totalRows step batchSize) {
+                val response = fetchRows(offset, batchSize, config)
+                val nextRootNode = parseResponse(response)
+                processRows(nextRootNode.path("rows"))
+            }
+
+            println("Successfully saved all articles for $config.")
+        } catch (ex: Exception) {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Create a custom exception

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+                .bodyToMono(String::class.java)
+                .retryWhen(
+                    Retry.fixedDelay(3, Duration.ofSeconds(5))
+                        .filter { throwable ->
+                            throwable is org.springframework.web.reactive.function.client.WebClientResponseException &&
+                                    throwable.statusCode.is5xxServerError
+                        }
+                )
+                .block() ?: throw RuntimeException("Failed to fetch rows")
+        } catch (ex: Exception) {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Create a custom exception

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+                processRows(nextRootNode.path("rows"))
+            }
+
+            println("Successfully saved all articles for $config.")
+        } catch (ex: Exception) {
+            println("Error fetching data for $config: ${ex.message}")
+        }
+    }
+
+    private fun fetchRows(offset: Int, length: Int, config: String): String {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Use a more descriptive data type for config, such as java.time
User a more descriptive name for the variable as well, as config is too generic and doesn't give any information that you are retrieving data related to a date range

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+                .retrieve()
+                .bodyToMono(String::class.java)
+                .retryWhen(
+                    Retry.fixedDelay(3, Duration.ofSeconds(5))
+                        .filter { throwable ->
+                            throwable is org.springframework.web.reactive.function.client.WebClientResponseException &&
+                                    throwable.statusCode.is5xxServerError
+                        }
+                )
+                .block() ?: throw RuntimeException("Failed to fetch rows")
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: throw custom exception

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+            throw ex
+        }
+    }
+
+    private fun parseResponse(response: String): JsonNode {
+        val objectMapper = ObjectMapper()
+        return objectMapper.readTree(response)
+    }
+
+    private fun processRows(rows: JsonNode) {
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt]: Refactor the title to something more descriptive.
What does process mean?

This method is inserting records in the database, so use something related to this action.

File: src/main/kotlin/sg/com/quantai/etl/services/BbcNewsTransformationService.kt [hunk 27]
```diff
@@ -0,0 +1,38 @@
+package sg.com.quantai.etl.services
+
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import sg.com.quantai.etl.data.BbcRawNewsArticle
+import sg.com.quantai.etl.data.BbcTransformedNewsArticle
```
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/services/BbcNewsTransformationService.kt]: Imported class does not exist

**Review by Ivsucram — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (karthikstar) [reply build.gradle]: fixed

**Comment by karthikstar** [comment]
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/transform   

To retrieve all transformed BBC news from database (within ETL):
curl -X GET http://localhost:10070/news_headlines/bbc/api/news_transformed

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_headlines/bbc_transformed/all

**Comment by karthikstar** [comment]
- Requested changes have been made
- Added unit tests for NewsHeadlinesController

Updated endpoints to check during functional test:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/transform

To retrieve all transformed BBC news from database (within ETL):
curl -X GET http://localhost:10070/news_headlines/bbc/api/news_transformed

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_headlines/bbc_transformed/all

File: src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt:1]: Line 9: 
The collection is still following the wrong name convention. It should be "news_articles_bbc," where "news_articles" denotes the scope that we intend to store, and "BBC" already denotes that it stores BBC raw data. This convention should also be described in the documentation.

Comment (Ivsucram) [reply src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt:1]: Line 10:
The data class name still follows the wrong name convention. It should be "NewsArticleBBC."

File: src/main/kotlin/sg/com/quantai/etl/repositories/BbcRawNewsRepository.kt
Comment (Ivsucram) [inline src/main/kotlin/sg/com/quantai/etl/repositories/BbcRawNewsRepository.kt:1]: The data class name still follows the wrong name convention. It should be "NewsArticleBBC."

File: src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+    @InjectMocks
+    private lateinit var newsHeadlinesController: NewsHeadlinesController
+
+    @BeforeEach
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
+    }
+
+    @Test
+    fun `fetchAndSaveNews should return OK status`() {
```
Comment (Ivsucram) [inline src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt]: The title of the test should be fully English formatted. So "should connect to fetch results" should suffice.

File: src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+        val response: ResponseEntity<String> = newsHeadlinesController.fetchAndSaveNews()
+
+        verify(bbcNewsApiService, times(1)).fetchAndSaveNewsArticlesForAllMonths()
+
+        assertEquals(HttpStatus.OK, response.statusCode)
+        assertEquals("News articles fetched and saved.", response.body)
+    }
+
+    @Test
+    fun `runTransformation should return OK when successful`() {
```
Comment (Ivsucram) [inline src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt]: The title of the test should be fully English formatted to describe which action and reaction is expected. So something like "Should apply transformations on BBC data" should suffice.

File: src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+        val response: ResponseEntity<String> = newsHeadlinesController.runTransformation()
+
+        verify(bbcNewsTransformationService, times(1)).transformAndSaveNewsArticles()
+
+        assertEquals(HttpStatus.OK, response.statusCode)
+        assertEquals("Transformation completed successfully and results have been stored in database.", response.body)
+    }
+
+    @Test
+    fun `runTransformation should return INTERNAL_SERVER_ERROR when transformation fails`() {
```
Comment (Ivsucram) [inline src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt]: The test title should be fully English formatted to describe which action and reaction is expected. So something like "Should yield proper exception if a transformation fails" or similar would be better.

File: src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+        doThrow(RuntimeException("Transformation failed")).`when`(bbcNewsTransformationService).transformAndSaveNewsArticles()
+
+        val response: ResponseEntity<String> = newsHeadlinesController.runTransformation()
+
+        assertEquals(HttpStatus.INTERNAL_SERVER_ERROR, response.statusCode)
+        assertEquals("Transformation failed: Transformation failed", response.body)
+    }
+
+    @Test
+    fun `getAllTransformedNews should return articles when available`() {
```
Comment (Ivsucram) [inline src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt]: The test title should be fully English formatted to describe which action and reaction is expected. I am not sure of the goal of this test by the title.

File: src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt [hunk 32]
```diff
@@ -0,0 +1,95 @@
...
+        val response: ResponseEntity<List<BbcTransformedNewsArticle>> = newsHeadlinesController.getAllTransformedNews()
+
+        verify(bbcTransformedNewsRepository, times(1)).findAll()
+
+        assertEquals(HttpStatus.OK, response.statusCode)
+        assertEquals(mockArticles, response.body)
+    }
+
+    @Test
+    fun `getAllTransformedNews should return NO_CONTENT when no articles are available`() {
```
Comment (Ivsucram) [inline src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt]: The test title should be fully English formatted to describe which action and reaction is expected. So something like "A NO_CONTENT response should be given if there are no content available" or something.

**Review by Ivsucram — CHANGES_REQUESTED** [review]

**Comment by karthikstar** [comment]
Updated: 
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_articles/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_articles/bbc/transform

To retrieve all transformed news from database (within ETL):
curl -X GET http://localhost:10070/news_articles/api/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_articles/all

**Comment by karthikstar** [comment]
- Requested code changes done.
- Fixed Unit Tests. All passed on my end

You can test the functionality using these updated endpoints:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_articles/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_articles/bbc/transform

To retrieve all transformed news from database (within ETL):
curl -X GET http://localhost:10070/news_articles/api/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_articles/all

**Comment by karthikstar** [comment]
Made code changes to:
- Skip fields 'author' and 'section'
- Check before inserting the data, in order to handle for multiple runs of fetch data

You can test the functionality using these updated endpoints:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_articles/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_articles/bbc/transform

To retrieve all transformed news from database (within ETL):
curl -X GET http://localhost:10070/news_articles/api/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_articles/all

**Pull request merged**